designs/node-repair.md

```diff
@@ -0,0 +1,93 @@
+#  Karpenter Node Auto Repair 
+
+## Problem Statement
+
+Karpenter users are observing that occasionally nodes can face failure mode that prevent them to be healthy in a Kubernetes clusters. These nodes stay in unhealthy state for a period of time without any actions, and in some cases forever with manual intervention. Karpenter continues to simulate pods to these nodes, but the Kubernetes scheduler won't schedule to them. This will leave pods unscheduled and cause downtime for applications. Customers have the need for karpenter to repair unhealthy nodes. The karpenter team has seen several failure mode such as disk pressure, memory pressure, networking outage etc. that can cause nodes to get into an unhealthy state Here are issues related to the topic:
```

Comment:
```suggestion
Nodes can experience failure modes that cause degredation to the underlying hardware, filesystem, or container environment. Some of these failure modes are surfaced through the Node object (consider referencing some of them) while others are not surfaced at all (consider referencing some of these). Node Problem Detector offers a way to surface these failures as additional status conditions on the node object.

In either case, even if a status condition is surfaced through the Node that indicates that the Node is unhealthy, Karpenter doesn't currently react to this unhealthiness in any way today. 
```

Just a suggestion on wording for this -- I'd consider framing this in how nodes experience failures today and which of these are currently surfaced -- along with what other components are needed to surface other types of failures. 

It might also be useful to show what this failure looks like on an actual node by showing status conditions for a node that is failing.

I'd also say that your comment on Karpenter continuing to schedule pods to these nodes may or may not be true. If the node gets tainted after initialization, we shouldn't be continuing to schedule there -- and the kube-scheduler won't either. I thought the larger issue was that we will just leave the node around even if it's unhealthy or even dead?